Rewrite as react-components-rails

renchap opened the pull request:
As discussed in #132, this is a rename and rewrite of this library.

Webpacker is now deprecated and wont get any new updates, but this library no longer relied on Webpacker.

The new name, `react-components-rails`, shows this (thanks @yubele for the suggestion).

I chose to start a new project nearly from scratch to have a much simpler and modern development setup for this. I copied the existing source code so it should behave the same way, but the Javascript library is now build using [`microbundle`](https://github.com/developit/microbundle) and written in Typescript.

TODO:
- [ ] add back UJS / Turbolinks / … support (but is it still current?)
- [ ] add back testing (using a single-file test app)
- [ ] add CI on Github Actions

renchap committed:
Rewrite as react-components-rails

yubele commented on README.md:
````diff
@@ -1,73 +1,41 @@
 
-```javascript
-//= require "turbolinks"
+ReactComponentsRails.setup({ Hello }) // ES6 shorthand for {Hello: Hello}
````
Since turbolink has been deprecated in rails7 in favor of turbo-rails, we have made it work as follows.

However, this method may affect performance.
```suggestion
document.addEventListener("turbo:load", function() {
  ReactComponentsRails.setup({ Hello }) // ES6 shorthand for {Hello: Hello}
})
```

net1957 commented:
in the gem react-components-rails (1.0.0.beta1) you miss the file react-components-rails.rb in lib. It should contain;
````
require "railtie"  if defined?(Rails)
require "helpers"
require "component"
````

without it, the Rails part is not loading.

and some unneeded files are in the gem:
![image](https://user-images.githubusercontent.com/17736/173255567-ea70c6e3-9a30-4b26-ad56-54dfa5374cd9.png)

net1957 commented:
during "compilation" with esbuild I get this error:
````
D:\Dvlt\ruby\projects\_active\Db_presenter>yarn build:dev
yarn run v1.22.18
$ esbuild app/javascript/application.js --bundle --sourcemap --inject:./app/javascript/react-shim.js --outdir=app/assets/builds
X [ERROR] Could not resolve "react-components-rails"

    app/javascript/application.js:15:33:
      15 │ import ReactComponentsRails from 'react-components-rails'
         ╵                                  ~~~~~~~~~~~~~~~~~~~~~~~~

  The module "./dist/react-components-rails.modern.js" was not found on the file system:

    node_modules/react-components-rails/package.json:15:15:
      15 │     "default": "./dist/react-components-rails.modern.js"
         ╵                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
````

I don't know microbundle, but as a workaround changing the default field to ./dist/react-components-rails.module.js
allow to get a running bundle and application.

my application package.json contains:
````
  "scripts": {
    "build:dev": "esbuild app/javascript/application.js --bundle --sourcemap --inject:./app/javascript/react-shim.js --outdir=app/assets/builds",
  }
````

renchap committed:
Switch from `microbundle` to `tsup` and from `yarn` to `pnpm`

renchap committed:
Support both the legacy and the new React 18+ API

renchap committed:
Release 1.0.0 beta 2

renchap commented:
Sorry, it took me quite a lot of time to come back to this project.

The gem should now be fixed, I reworked the JS build process and added support for the new React 18+ render API.
This makes a dynamic import for `react-dom/client` to test if you are running React 18+ or not, using `await import("react-dom/client")`. Bundlers should be able to transform this as needed, but I suspect this might break in some setups.

I just published `react-components-rails` 1.0.0-beta2 to both NPM and Rubygems.

I plan to merge this to `main` before the end of the month, and then add back some tests before the final release.

renchap committed:
Remove dependency on lodash

renchap committed:
Remove `async` functions

This avoids the need to call `regenerator-runtime` when transpiled

renchap committed:
Better exports definition

renchap committed:
Switch to a proper class

renchap committed:
Release 1.0.0 beta 3

renchap committed:
Use `.default` is the bundler imports `react-dom` in it

Also use the direct import from `react-dom` as fallback, not our own variable

renchap committed:
Switch back to an ESM package

renchap committed:
Release 1.0.0 beta 4

renchap commented:
I tested the latest beta (`beta4`) with both Webpacker and Vite-Ruby.

Lodash is also removed, this package no longer has any dependency :)

net1957 commented:
It's on production for some times now without any problem :)

my configuration at this time:
- Rails 7.04
- React 18.2.0
- react-components-rails 1.0.0-beta.4
- esbuild 0.15.16

Can we expect a new non beta release?

rossta commented:
We are also successfully building with 1.0.0-beta.4 on both React 17 (with the my [documented webpack workaround](https://github.com/renchap/webpacker-react/issues/147)) and React 18 on webpack 5. It would be great to see this merged to `main` with a 1.0.0 release.

rossta commented:
@renchap I notice that ujs integration hasn't yet been implemented. Would you like some assistance completing this branch?
